airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/report_streams.py

```diff
@@ -120,37 +129,43 @@ def read_records(
         report_date = stream_slice[self.cursor_field]
         report_infos = self._init_reports(report_date)
         logger.info(f"Waiting for {len(report_infos)} report(s) to be generated")
-        # According to Amazon Ads API docs metric generation takes maximum 15
-        # minutes. But in case reports wont be generated we dont want this stream to
-        # hung forever. Store timepoint when report generation has started to
-        # check if it takes to long to break a loop.
-        start_time_point = datetime.now()
-        while report_infos and datetime.now() <= start_time_point + self.REPORT_WAIT_TIMEOUT:
-            completed_reports = []
-            logger.info(f"Checking report status, {len(report_infos)} report(s) remained")
-            for report_info in report_infos:
-                report_status, download_url = self._check_status(report_info)
-                if report_status == Status.FAILURE:
-                    raise Exception(f"Report for {report_info.profile_id} with {report_info.record_type} type generation failed")
-                elif report_status == Status.SUCCESS:
-                    metric_objects = self._download_report(report_info, download_url)
-                    for metric_object in metric_objects:
-                        yield self._model(
-                            profileId=report_info.profile_id,
-                            recordType=report_info.record_type,
-                            reportDate=report_date,
-                            metric=metric_object,
-                        ).dict()
-                    completed_reports.append(report_info)
-            for completed_report in completed_reports:
-                report_infos.remove(completed_report)
-            if report_infos:
-                logger.info(f"{len(report_infos)} report(s) remained, taking {self.CHECK_INTERVAL_SECONDS} seconds timeout")
-                time.sleep(self.CHECK_INTERVAL_SECONDS)
-        if not report_infos:
-            logger.info("All reports have been processed")
-        else:
-            raise Exception("Not all reports has been processed due to timeout")
+
+        self._try_read_records(report_infos)
+
+        for report_info in report_infos:
+            for metric_object in report_info.metric_objects:
+                yield self._model(
+                    profileId=report_info.profile_id,
+                    recordType=report_info.record_type,
+                    reportDate=report_date,
+                    metric=metric_object,
+                ).dict()
+
+    @backoff.on_exception(
+        backoff.constant,
+        (ReportGenerationFailure, ReportGenerationInProgress),
+        max_time=REPORT_WAIT_TIMEOUT,
+    )
+    def _try_read_records(self, report_infos):
```

Comment:
Good job refactoring this messy code. But you have changed logic so now it wouldn't wait for reports status update and expect it to be completed immediately? You misunderstood REPORT_WAIT_TIMEOUT parameter, this is not time required for downloading report, its maximum time that requires for generating async report.  Please read comment that you have deleted :) 

I agree on that if single report failed with 500 error it should be retried but in this case we should take into account report processing  time for each report. 
Wouldn't you mind if I pick up this PR and continue work for you. @sherifnada  what do you think on priority of this ticket?

Reply:
> But you have changed logic so now it wouldn't wait for reports status update and expect it to be completed immediately?

As I wrote in the PR description above, I refactored the polling generator as a decorated function. As a result, it polls report status for up to `REPORT_WAIT_TIMEOUT`, i.e., 20 minutes (see the [documentation](https://github.com/litl/backoff) for `max_time` property).

> You misunderstood REPORT_WAIT_TIMEOUT parameter, this is not time required for downloading report, its maximum time that requires for generating async report.

False; see the comment above.

> I agree on that if single report failed with 500 error it should be retried but in this case we should take into account report processing time for each report.

As already does yours, my implementation uses the same timeout for an entire report batch stored in `report_infos`. Report generation is being initiated simultaneously for the batch and, therefore, should be completed in the same 20 minutes window. Even if they're generated not in the same order as initiated, the polling function takes and processes whichever is generated first.

> Wouldn't you mind if I pick up this PR and continue work for you. @sherifnada what do you think on priority of this ticket?

Of course not, that would be great 😊 I started this PR because my original issue #6767 hasn't got any attention yet.

To add more context to this issue. Amazon Ads connector is virtually unusable with a real account with several actively running campaigns. Report extraction fails due to one or another error. Roughly 1 in 20, if not less, sync succeeds.

Reply:
@monai Sorry for long time no response, lost this conversation.  Thanks for detailed answer, it makes sense, lets merge it.